monai/_extensions/hausdorff/hausdorff_cuda.cu

```diff
@@ -0,0 +1,2610 @@
+
+
+    return fbArgs;
+
+}
+
+
+
+#pragma once
+template <typename T>
+ForBoolKernelArgs<T> mainKernelsRun(ForFullBoolPrepArgs<T>& fFArgs, const int WIDTH, const int HEIGHT, const int DEPTH, cudaStream_t stream, bool resToSave = false,bool  resIterNeeded=false
+) {
+
+    //cudaDeviceReset();
+    cudaError_t syncErr;
+    cudaError_t asyncErr;
+
+    occupancyCalcData occData = getOccupancy<T>();
+
+    //pointers ...
+    ForBoolKernelArgs<T> fbArgs = executeHausdoff(fFArgs, WIDTH, HEIGHT, DEPTH, occData, resToSave, stream,1.0, resIterNeeded);
+
+    checkCuda(cudaDeviceSynchronize(), "last ");
+
+    /////////// error handling
+    syncErr = cudaGetLastError();
+    asyncErr = cudaDeviceSynchronize();
+    if (syncErr != cudaSuccess) printf("Error in syncErr: %s\n", cudaGetErrorString(syncErr));
+    if (asyncErr != cudaSuccess) printf("Error in asyncErr: %s\n", cudaGetErrorString(asyncErr));
+
+
+    cudaDeviceReset();
+
+    return fbArgs;
+}
+
+
+
+
+//    modified  from https://github.com/pytorch/pytorch/blob/61d6c4386459441710fb4cfa2929a3f77e95e5f7/aten/src/ATen/Dispatch.h
+#define AT_DISPATCH_ALL_TYPESWithBool(TYPE, NAME, ...)                               \
+  [&] {                                                                      \
+    const auto& the_type = TYPE;                                             \
+    /* don't use TYPE again in case it is an expensive or side-effect op  */ \
+    at::ScalarType _st = ::detail::scalar_type(the_type);                    \
+    RECORD_KERNEL_FUNCTION_DTYPE(NAME, _st);                                 \
+    switch (_st) {                                                              \
+      AT_PRIVATE_CASE_TYPE(NAME, at::ScalarType::Byte, uint8_t, __VA_ARGS__)    \
+      AT_PRIVATE_CASE_TYPE(NAME, at::ScalarType::Char, int8_t, __VA_ARGS__)     \
+      AT_PRIVATE_CASE_TYPE(NAME, at::ScalarType::Double, double, __VA_ARGS__)   \
+      AT_PRIVATE_CASE_TYPE(NAME, at::ScalarType::Float, float, __VA_ARGS__)     \
+      AT_PRIVATE_CASE_TYPE(NAME, at::ScalarType::Int, int32_t, __VA_ARGS__)     \
+      AT_PRIVATE_CASE_TYPE(NAME, at::ScalarType::Long, int64_t, __VA_ARGS__)    \
+      AT_PRIVATE_CASE_TYPE(NAME, at::ScalarType::Short, int16_t, __VA_ARGS__)   \
+      AT_PRIVATE_CASE_TYPE(NAME, at::ScalarType::Bool, bool, __VA_ARGS__)   \
+      default:                                                                  \
+        AT_ERROR(#NAME, " not implemented for '", toString(_st), "'");          \
+    }                                                                           \
+  }()
+
+
+
+
+
+template <typename T>
+int getHausdorffDistance_CUDA_Generic(at::Tensor goldStandard,
+    at::Tensor algoOutput
+    , int WIDTH, int HEIGHT, int DEPTH, float robustnessPercent, bool resIterNeeded, at::Tensor numberToLookFor) {
+    //TODO() use https ://pytorch.org/cppdocs/notes/tensor_cuda_stream.html
+    cudaStream_t stream1;
+    cudaStreamCreate(&stream1);
```

Comment:
Can we create 2 streams for readability, memCpyStream and kernelStream?

Reply:
Frankly I do not understand the idea (forgive me for my ignorance) but is memory allocation would be done on diffrent stream than kernel code would it not lead to synchronization problems?

Reply:
That appears to be what you are doing. The kenral invocations aren't provided a stream and so will be on the default stream. You then use the stream allocated here in various mem calls (copy, free, aloc). This is a valid optimisation technique aslong as the operations are indeed parallelizable.

Reply:
ok true, thanks  !